#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""

## ChIPseq-pipeline

+ 1. align IP/Input to genome

```
bowtie2 -x {genome} -U {fq} -q {options} --very-sensitive-local -p {cpu} > {out.bam}
```

+ 2. call peaks using MACS32 with --SPMR

```
macs2_broad_opt="--braod"
keep_dup_option='--keep-dup auto'

macs2 callpeak -f BAM -t {IP.bam} -c {Input.bam} -g {gsize} {macs2_broad_opt} --outdir {output} -n {prefix} -B --SPMR {keep_dup_option}
macs2 bdgcmp -t ${treat_pileup.bdg} -c {control_lambda.bdg} -o {ppois.bdg} -m ppois
macs2 bdgcmp -t ${treat_pileup.bdg} -c {control_lambda.bdg} -o {FE.bdg} 
macs2 bdgcmp -t ${treat_pileup.bdg} -c {control_lambda.bdg} -o {LogLR.bdg} -m logLR -p 0.00001  
```


read the depth, parse file: {prefix}.clean_peaks.xls

```
effective_depth_ip="tags after filtering in treatment"
depth_ip="tags in treatment"
NormScaleIP="1e6/{depth_ip}"

effective_depth_input="after filtering in control"
depth_input="tags in control"
NormScaleInput="1e6/{depth_input"
```

Example of the file:

```
# This file is generated by MACS version 2.1.1.20160309
# Command line: callpeak -f BAM -t {IP.bam} -c {Input.bam} -g 139741753 --broad --outdir macs2_peaks_calling -n {prefix} -B --SPMR
# ARGUMENTS LIST:
# name = 97FL_Nxf2MT_H3K9me3_rep1_1.clean
# format = BAM
# ChIP-seq file = ['genome_mapping_unique_only/97FL_Nxf2MT_H3K9me3_rep1_1.clean.dm3.IP.sorted.bam']
# control file = ['genome_mapping_unique_only/97FL_Nxf2MT_H3K9me3_rep1_1.clean.dm3.Input.sorted.bam']
# effective genome size = 1.40e+08
# band width = 300
# model fold = [5, 50]
# qvalue cutoff for narrow/strong regions = 5.00e-02
# qvalue cutoff for broad/weak regions = 1.00e-01
# Larger dataset will be scaled towards smaller dataset.
# Range for calculating regional lambda is: 1000 bps and 10000 bps
# Broad region calling is on
# Paired-End mode is off
# MACS will save fragment pileup signal per million reads

# tag size is determined as 100 bps
# total tags in treatment: 7978071
# tags after filtering in treatment: 2384854
# maximum duplicate tags at the same position in treatment = 1
# Redundant rate in treatment: 0.70
# total tags in control: 10555283
# tags after filtering in control: 6639591
# maximum duplicate tags at the same position in control = 1
# Redundant rate in control: 0.37
# d = 122
# alternative fragment length(s) may be 122 bps
```









+ 3. direct aligement

alignment IP/Input reads

```
bowtie2 -x {genome} -1 {fq1} -2 {fq2} -q {bowtie2PhredOption} -a -X 800 --no-mixed --quiet
```

make summary graph

```
grep -v 'NM_' {tr_size} | grep -v 'NR_' | grep -v 'FBtr' > {te_size}

bedtools genomecov -i {IP.unique.bed} -g {tr_size} -bg -scale {IP_scale} > {IP.unique.bg}
bedGraphToBigWig ${IP.unique.bg} {tr_size} {IP.unique.bw}

bedtools genomecov -i {Input.unique.bed} -g {tr_size} -bg -scale {Input_scale} > {Input.unique.bg}
bedGraphToBigWig ${Input.unique.bg} {tr_size} {Input.unique.bw}

# 
awk -v bgP=${bgP} -v bgM=${bgM} -v binSize=${BINSIZE} '{print "bigWigSummary", bgP, $1, 0, $2, $2, "| sed -e \x27s/n\\/a/0/g\x27 >", bgP"."$1; print "bigWigSummary", bgM, $1, 0, $2, $2, "| sed -e \x27s/n\\/a/0/g\x27 >", bgM"."$1;}' ${dDIRECTMAPPING_DIR}/transposon.sizes > $paraFile 
```




```
Alignment mode:
0, unique_only
1. unique_only
2. randomely assigned multimapper
4. CSEM allocated multimapper
 
"""


