---
title: "Quality Control Report"
date: "`r Sys.Date()`"
output:
  prettydoc::html_pretty:
    theme: hpstr
    toc: true
    toc_depth: 2
    highlight: github
    df_print: kable
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r global_options, include=FALSE}
library(goldclipReport)
library(fastqcr)
library(dplyr)
library(ggplot2)
library(cowplot)
library(knitr)
suppressPackageStartupMessages(library(kableExtra))

knitr::opts_chunk$set(fig.width = 12, 
                      fig.height = 8, 
                      fig.path = 'Figs/',
                      echo = FALSE,
                      cache = FALSE,
                      prompt = FALSE,
                      tidy = FALSE,
                      comment = NA,
                      message = FALSE,
                      warning = FALSE,
                      rownames.print = FALSE)
options(width=150)

print_df <- function(x){
  x %>%
  kableExtra::kable() %>%
  kableExtra::kable_styling() %>%
  kableExtra::scroll_box(width = "100%", height = "400px")
}

qc.path <- "/home/wangming/work/devel_pipeline/hipipe/rnaseq/results/report/qc"
```


## 1. Summary

```{r fig_qc, fig.align = "middle", fig.cap = "overview of analysis"}
qc <- qc_aggregate(qc.path, progressbar = FALSE)
qcgs <- qc_stats(qc)
qcgs <- dplyr::mutate(qcgs, 
               tot.seq = as.numeric(tot.seq) / 1e6,
               pct.dup = paste0(pct.dup, '%'),
               pct.gc  = paste0(pct.gc, '%')) %>%
  dplyr::select(sample, tot.seq, pct.dup, pct.gc, seq.length)
names(qcgs) <- c('Sample', 'Seqs (Million)', '% Dup', '% GC', 'Length (nt)')
print_df(qcgs)
```


## 2. Distribution of Base Quality

> From Illumina:

Q scores are defined as a property that is logarithmically related to the base calling error probabilities (P)

$$Q = - 10 log_{10} P $$  

For example, if Phred assigns a Q score of 30 (Q30) to a base, this is equivalent to the probability of an incorrect base call 1 in 1000 times (Table 1). This means that the base call accuracy (i.e., the probability of a correct base call) is 99.9%. A lower base call accuracy of 99% (Q20) will have an incorrect base call probability of 1 in 100, meaning that every 100 bp sequencing read will likely contain an error. When sequencing quality reaches Q30, virtually all of the reads will be perfect, having zero errors and ambiguities. This is why Q30 is considered a benchmark for quality in next-generation sequencing. By comparison, Sanger sequencing systems generally produce base call accuracy of ~99.4%, or $Q20^{3}$. Low Q scores can increase false-positive variant calls, which can result in inaccurate conclusions and higher costs for validation experiments ^[Illumina, Inc, https://www.illumina.com/documents/products/technotes/technote_Q-Scores.pdf]. 


```{r table_quality, }

qual <- data.frame("Score" = as.character(c(10, 20, 30, 40, 50)),
                   "Incorrect" = c("1 in 10", 
                                   "1 in 100",
                                   "1 in 1000", 
                                   "1 in 10000",
                                   "1 in 100000"),
                   "Accuracy" = c("90%", "99%", "99.9%", "99.99%", "99.999%"),
                   stringsAsFactors = FALSE)
print_df(qual)

```

Score = "Phred Quality Score", Incorrect = "Probability of Incorrect Base Call", "Accuracy" = "Base Call Accuracy".


```{r fig_quality}
## grid plots
qc.files <- FastqcFiles(qc.path)

for(x in qc.files) {
  p <- fastqc_plot(x)
  print(p)
}
```

## 3. Quality Control Report in details


```{r}
qc.html <- list.files(qc.path, "*_fastqc.html", all.files = TRUE, 
                      full.names = TRUE, recursive = FALSE)
qc.url <- glue::glue("[{qc.html}](./{qc.html})")
qc.url
```











